src/ai/NetworkInterface.js

```diff
@@ -0,0 +1,136 @@
+                                //  ALSO NEED TO CONSIDER: if someone in the session already logged in with ai, people who joined later how do they know how to join?
+
+                                // todo Steve: currently, one edge case: when a user later join the session, before subscribing all the avatars & get the ai api keys,
+                                //  they input another ai api key. This way, even later users might get either api keys, maybe activating 2 different kinds of azure gpt instances
+                                //  solution: need to store this info in the session storage, and once set, don't update it. This way later user will get this info faster, and cannot modify it
+                                // console.log(`Broadcast endpoint and apikey to other avatars: ${endpoint}, ${azureApiKey}`);
+                                realityEditor.avatar.network.sendAiApiKeys(realityEditor.avatar.getMyAvatarNodeInfo(), {
+                                    endpoint: endpoint,
+                                    azureApiKey: azureApiKey,
+                                });
+                            }
+                            resolve('success');
+                        } else {
+                            reject(`Endpoint didn't successfully init AI API keys: ${endpoint}, ${azureApiKey}: ${res.answer}`);
+                        }
+                    }
+                });
+        });
+    }
+
+    async postFunctionResultToAI(functionCallId, functionResult) {
+        let route = '/ai/continue-query';
+
+        let worldId = realityEditor.worldObjects.getBestWorldObject();
+        let ip = worldId.ip;
+        let port = realityEditor.network.getPort(worldId);
+
+        return new Promise((resolve, reject) => {
+
+            realityEditor.network.postData(realityEditor.network.getURL(ip, port, route),
+                { functionCallId, result: functionResult },
+                function (err, res) {
+                    // console.log(res);
+                    if (err) {
+                        console.warn('ai/continue-query error:', err);
+                        reject(err);
+                    } else {
+                        resolve(res);
+                    }
+                });
+
+        });
+    }
+
+    async processToolStateWithPrompts(applicationId, state, prompts) {
+        let route = '/ai/process-state-with-prompts';
+
+        // let worldId = await this.findWorldId();
+        let worldId = realityEditor.worldObjects.getBestWorldObject();
+        let ip = worldId.ip;
+        let port = realityEditor.network.getPort(worldId);
+
+        return new Promise((resolve, reject) => {
+            realityEditor.network.postData(realityEditor.network.getURL(ip, port, route),
+                {
+                    applicationId,
+                    state,
```

Comment:
I don't see the edge-server routes supporting this feature, but just want to make sure that the state is being sent as part of the system message and not an actual message in the chat history. This ensures we don't send earlier states over and over again, using up tons of tokens each time we send the chat history as context.

Reply:
I will look into this. Is there a difference between system messages and chat history messages in terms of how they get used as tokens? I think I've been sending the system messages in every time the user types a new input, even though some of them don't change.